Mathlib/CategoryTheory/Limits/FormalCoproducts.lean

```diff
@@ -0,0 +1,429 @@
+@[simps!] def isoOfComponents {X Y : FormalCoproduct.{w} C} (e : X.I ≃ Y.I)
+    (h : ∀ i, X.obj i ≅ Y.obj (e i)) : X ≅ Y where
+  hom := { f := e, φ := fun i ↦ (h i).hom }
+  inv := { f := e.symm, φ := fun i ↦ eqToHom (by simp) ≫ (h (e.symm i)).inv }
+  hom_inv_id := by ext <;> aesop
+  inv_hom_id := by ext <;> aesop
+
+variable (C) in
+/-- An object of the original category produces a formal coproduct on that object only, so indexed
+by `PUnit`, the type with one element. -/
+@[simps!] def incl : C ⥤ FormalCoproduct.{w} C where
+  obj X := ⟨PUnit, fun _ ↦ X⟩
+  map f := ⟨fun _ ↦ PUnit.unit, fun _ ↦ f⟩
+
+section fromIncl
+
+variable {X : C} {Y : FormalCoproduct.{w} C}
+
+/-- A map `incl(X) ⟶ Y` is specified by an element of `Y`'s indexing set, and then a morphism
+`X ⟶ Y.obj i` in the original category. -/
+@[simps!] def Hom.fromIncl (i : Y.I) (f : X ⟶ Y.obj i) : (incl C).obj X ⟶ Y :=
+  ⟨fun _ ↦ i, fun _ ↦ f⟩
+
+/-- A map `incl(X) ⟶ Y` is specified by an element of `Y`'s indexing set, and then a morphism
+`X ⟶ Y.obj i` in the original category. -/
+def Hom.asSigma (f : (incl C).obj X ⟶ Y) : Σ (i : Y.I), X ⟶ Y.obj i :=
+  ⟨f.f PUnit.unit, f.φ PUnit.unit⟩
+
+lemma Hom.fromIncl_asSigma (f : (incl C).obj X ⟶ Y) :
+    Hom.fromIncl f.asSigma.fst f.asSigma.snd = f := by
+  ext <;> aesop
+
+end fromIncl
+
+-- This is probably some form of adjunction?
+/-- A map `incl(X) ⟶ Y` is specified by an element of `Y`'s indexing set, and then a morphism
+`X ⟶ Y.obj i` in the original category. -/
+@[simps!] def inclHomEquiv (X : C) (Y : FormalCoproduct.{w} C) :
+    ((incl C).obj X ⟶ Y) ≃ (i : Y.I) × (X ⟶ Y.obj i) where
+  toFun f := f.asSigma
+  invFun f := .fromIncl f.1 f.2
+  left_inv f := f.fromIncl_asSigma
+  right_inv _ := rfl
+
+/-- `incl` is fully faithful, which means that `(X ⟶ Y) ≃ (incl(X) ⟶ incl(Y))`. -/
+@[simps!] def fullyFaithfulIncl : (incl C).FullyFaithful where
+  preimage f := f.φ PUnit.unit
+
+instance : (incl C).Full :=
+  fullyFaithfulIncl.full
+
+instance : (incl C).Faithful :=
+  fullyFaithfulIncl.faithful
+
+/-- A family of maps with the same target can be turned into one arrow in the category of formal
+coproducts. This is used in Čech cohomology. -/
+@[simps!] def homOfPiHom (X : C) {J : Type w} (f : (j : J) → C) (φ : (j : J) → f j ⟶ X) :
+    FormalCoproduct.mk _ f ⟶ (incl C).obj X :=
+  ⟨fun _ ↦ PUnit.unit, φ⟩
+
+
+section Coproduct
+
+variable (𝒜 : Type w) (f : 𝒜 → FormalCoproduct.{w} C) (t X : FormalCoproduct.{w} C)
+
+/-- We construct explicitly the data that specify the coproduct of a given family of formal
+coproducts. -/
+def cofan : Cofan f :=
+  Cofan.mk ⟨(i : 𝒜) × (f i).I, fun p ↦ (f p.1).obj p.2⟩
+    fun i ↦ ⟨fun x ↦ ⟨i, x⟩, fun x ↦ 𝟙 ((f i).obj x)⟩
+
+section simp_lemmas
+
+variable {𝒜 f}
+
+theorem cofan_inj (i : 𝒜) : (cofan 𝒜 f).inj i = ⟨fun x ↦ ⟨i, x⟩, fun x ↦ 𝟙 ((f i).obj x)⟩ := rfl
+-- JH: this is probably a bad theorem? At least not good for simp
```

Comment:
Not good for simp, but I don't think it's a problem to keep it as an ordinary lemma.
```suggestion
```